crates/apollo-compiler/src/name.rs

```diff
@@ -0,0 +1,511 @@
+    /// or the [`name!`][crate::name] macro, return the string with `'static` lifetime.
+    pub fn as_static_str(&self) -> Option<&'static str> {
+        if self.tagged_file_id.tag() == TAG_STATIC {
+            let raw_slice = NonNull::slice_from_raw_parts(self.ptr, self.len());
+            // SAFETY: the tag indicates `self.ptr` came from `Self::ptr_and_tag_from_static`,
+            // so it has the static lifetime and points to valid UTF-8 of the correct length.
+            Some(unsafe { std::str::from_utf8_unchecked(raw_slice.as_ref()) })
+        } else {
+            None
+        }
+    }
+
+    fn as_arc(&self) -> Option<ManuallyDrop<Arc<str>>> {
+        if self.tagged_file_id.tag() == TAG_ARC {
+            let raw_slice = NonNull::slice_from_raw_parts(self.ptr, self.len())
+                .as_ptr()
+                .cast_const();
+
+            // SAFETY:
+            //
+            // * The tag indicates `self.ptr` came from `Arc::into_raw` in `ptr_and_tag_with_arc`
+            // * `Arc::from_raw` normally moves ownership away from the raw pointer,
+            //   `ManuallyDrop` counteracts that
+            Some(ManuallyDrop::new(unsafe {
+                Arc::from_raw(raw_slice as *const str)
+            }))
+        } else {
+            None
+        }
+    }
+
+    /// If this `Name` was created with [`new_static`][Self::new_static]
+    /// or the [`name!`][crate::name] macro, return the string with `'static` lifetime.
+    ///
+    /// Otherwise, return a clone of the `Arc` used internally for this `Name`.
+    pub fn to_static_str_or_cloned_arc(&self) -> Result<&'static str, Arc<str>> {
+        self.as_static_str().ok_or_else(|| {
+            let manually_drop = self.as_arc().unwrap();
+            Arc::clone(&manually_drop)
+        })
+    }
+
+    /// Returns whether the given string is a valid GraphQL name.
+    ///
+    /// <https://spec.graphql.org/October2021/#Name>
+    pub const fn valid_syntax(value: &str) -> bool {
+        let bytes = value.as_bytes();
+        let Some(&first) = bytes.first() else {
+            return false;
+        };
+        if !Self::char_is_name_start(first) {
+            return false;
+        }
+        // TODO: iterator when available in const
+        let mut i = 1;
+        while i < bytes.len() {
+            if !Self::char_is_name_continue(bytes[i]) {
+                return false;
+            }
+            i += 1
+        }
+        true
+    }
+
+    fn check_valid_syntax(
+        value: &str,
+        location: Option<NodeLocation>,
+    ) -> Result<(), InvalidNameError> {
+        if Self::valid_syntax(value) {
+            Ok(())
+        } else {
+            Err(InvalidNameError {
+                name: value.to_owned(),
+                location,
+            })
+        }
+    }
+
+    /// <https://spec.graphql.org/October2021/#NameStart>
+    const fn char_is_name_start(byte: u8) -> bool {
```

Comment:
```suggestion
    const fn is_char_name_start(byte: u8) -> bool {
```

Reply:
`char` (should be `byte` since i changed the argument type) is the subject of `is` not the object. I went with just `is_name_start`, it’s a private function and the type of argument is already apparent in the rest of the signature.